### ISSUE TITLE
Docker stack recreating service randomly if more than one secret is specified

### ISSUE THREAD

**Issue by riaan53** [moby/moby#33945]
**Description**

When specifying more than one secret for a service in the compose file and using docker stack to update the service it randomly recreates the service. It looks like the same issue that was fixed for the environmental variables in https://github.com/moby/moby/pull/32364 but for secrets.

**Steps to reproduce the issue:**
1. Create some secrets
echo "This is a secret" | docker secret create secret1 -
echo "This is a secret" | docker secret create secret2 -
echo "This is a secret" | docker secret create secret3 -

2. Create compose file
```yml
version: '3.3'

services:
  redis:
    image: redis:alpine
    secrets:
      - secret1
      - secret2
      - secret3

secrets:
  secret1:
    external: true
  secret2:
    external: true
  secret3:
    external: true
```
3. Create stack
docker stack deploy --compose-file docker-compose.yml test

4. Update stack a few times and note the create time with docker ps
docker stack deploy --compose-file docker-compose.yml test
docker stack deploy --compose-file docker-compose.yml test
docker stack deploy --compose-file docker-compose.yml test
docker stack deploy --compose-file docker-compose.yml test
docker stack deploy --compose-file docker-compose.yml test

**Describe the results you received:**
Service is recreated randomly. Note that when changing the compose file to only one secret that service is never recreated.

**Describe the results you expected:**
Service is not recreated if nothing changed.

**Output of `docker version`:**
```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:31:53 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:51:55 2017
 OS/Arch:      linux/amd64
 Experimental: true
```
**Output of `docker info`:**
```
Containers: 5
 Running: 1
 Paused: 0
 Stopped: 4
Images: 1
Server Version: 17.06.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: 1s53liygwu73cgkkrwpb11pbw
 Is Manager: true
 ClusterID: v34ixvlpqlsji69gw41l5gmmr
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: 192.168.65.2
 Manager Addresses:
  192.168.65.2:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.31-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 9.744GiB
Name: moby
ID: XCBJ:EEN2:NDQK:TZCC:FOZV:L35K:KCMF:3QWG:7IYF:VTIT:6LOE:3TPD
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 44
 Goroutines: 169
 System Time: 2017-07-04T12:45:52.082943373Z
 EventsListeners: 2
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```
Docker for mac used in this example but also seen on Ubuntu 16.04 LTS.

**Comment by riaan53** [comment]
Some more info in this issue. From the container spec below you can see that the secrets are in different order after updating causing the stack update to re-deploy the container.

```json
[
    {
        "ID": "w8707bxxil097nhhbyimy3fdw",
        "Version": {
            "Index": 13417
        },
        "CreatedAt": "2017-07-11T08:46:26.179500183Z",
        "UpdatedAt": "2017-07-11T08:48:11.364973919Z",
        "Spec": {
            "Name": "test_redis",
            "Labels": {
                "com.docker.stack.image": "redis:alpine",
                "com.docker.stack.namespace": "test"
            },
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "redis:alpine@sha256:2b1600c032e7653d079e9bd1eb80df5c99733795691f5ae9bca451bec325b7ea",
                    "Labels": {
                        "com.docker.stack.namespace": "test"
                    },
                    "Privileges": {
                        "CredentialSpec": null,
                        "SELinuxContext": null
                    },
                    "StopGracePeriod": 10000000000,
                    "DNSConfig": {},
                    "Secrets": [
                        {
                            "File": {
                                "Name": "secret2",
                                "UID": "0",
                                "GID": "0",
                                "Mode": 292
                            },
                            "SecretID": "xm6y43umtbpo684jluek86zxy",
                            "SecretName": "secret2"
                        },
                        {
                            "File": {
                                "Name": "secret3",
                                "UID": "0",
                                "GID": "0",
                                "Mode": 292
                            },
                            "SecretID": "l4ghp3tcmg8ed1oo8tma66qz0",
                            "SecretName": "secret3"
                        },
                        {
                            "File": {
                                "Name": "secret1",
                                "UID": "0",
                                "GID": "0",
                                "Mode": 292
                            },
                            "SecretID": "yebas4utmr2w3dqiuu3840f3d",
                            "SecretName": "secret1"
                        }
                    ]
                },
                "Resources": {},
                "RestartPolicy": {
                    "Condition": "any",
                    "Delay": 5000000000,
                    "MaxAttempts": 0
                },
                "Placement": {
                    "Platforms": [
                        {
                            "Architecture": "amd64",
                            "OS": "linux"
                        }
                    ]
                },
                "Networks": [
                    {
                        "Target": "m8b14ud16viwfgs8mttr5djn0",
                        "Aliases": [
                            "redis"
                        ]
                    }
                ],
                "ForceUpdate": 0,
                "Runtime": "container"
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "Monitor": 5000000000,
                "MaxFailureRatio": 0,
                "Order": "stop-first"
            },
            "RollbackConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "Monitor": 5000000000,
                "MaxFailureRatio": 0,
                "Order": "stop-first"
            },
            "EndpointSpec": {
                "Mode": "vip"
            }
        },
        "PreviousSpec": {
            "Name": "test_redis",
            "Labels": {
                "com.docker.stack.image": "redis:alpine",
                "com.docker.stack.namespace": "test"
            },
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "redis:alpine@sha256:2b1600c032e7653d079e9bd1eb80df5c99733795691f5ae9bca451bec325b7ea",
                    "Labels": {
                        "com.docker.stack.namespace": "test"
                    },
                    "Privileges": {
                        "CredentialSpec": null,
                        "SELinuxContext": null
                    },
                    "Secrets": [
                        {
                            "File": {
                                "Name": "secret1",
                                "UID": "0",
                                "GID": "0",
                                "Mode": 292
                            },
                            "SecretID": "yebas4utmr2w3dqiuu3840f3d",
                            "SecretName": "secret1"
                        },
                        {
                            "File": {
                                "Name": "secret2",
                                "UID": "0",
                                "GID": "0",
                                "Mode": 292
                            },
                            "SecretID": "xm6y43umtbpo684jluek86zxy",
                            "SecretName": "secret2"
                        },
                        {
                            "File": {
                                "Name": "secret3",
                                "UID": "0",
                                "GID": "0",
                                "Mode": 292
                            },
                            "SecretID": "l4ghp3tcmg8ed1oo8tma66qz0",
                            "SecretName": "secret3"
                        }
                    ]
                },
                "Resources": {},
                "Placement": {
                    "Platforms": [
                        {
                            "Architecture": "amd64",
                            "OS": "linux"
                        }
                    ]
                },
                "Networks": [
                    {
                        "Target": "m8b14ud16viwfgs8mttr5djn0",
                        "Aliases": [
                            "redis"
                        ]
                    }
                ],
                "ForceUpdate": 0,
                "Runtime": "container"
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "EndpointSpec": {
                "Mode": "vip"
            }
        },
        "Endpoint": {
            "Spec": {
                "Mode": "vip"
            },
            "VirtualIPs": [
                {
                    "NetworkID": "m8b14ud16viwfgs8mttr5djn0",
                    "Addr": "10.0.2.2/24"
                }
            ]
        },
        "UpdateStatus": {
            "State": "completed",
            "StartedAt": "2017-07-11T08:47:58.421593714Z",
            "CompletedAt": "2017-07-11T08:48:11.364942771Z",
            "Message": "update completed"
        }
    }
]
```

**Comment by cpuguy83** [comment]
Thanks, I suspect the CLI is not sorting the secrets.

**Comment by thaJeztah** [comment]
Looks like a duplicate of https://github.com/moby/moby/issues/34746, which is fixed in Docker 17.09 and up through https://github.com/docker/cli/pull/509. Docker EE 17.06 also has that fix
